src/integrationtests/java/com/aws/greengrass/integrationtests/certificateauthority/OfflineAuthenticationTest.java

```diff
@@ -276,4 +280,32 @@ void GIVEN_storedCertificates_WHEN_refreshEnabled_THEN_storedCertificatesRefresh
        // This one should have been removed given it is no longer attached
        assertNull(thingB);
    }
+
+   @Test
+    void GIVEN_clientConnectsWhileOnline_WHEN_offlineAndCertificateRevoked_THEN_backOnlineAndClientRejected() throws
+           Exception {
+        // Given
+       network.goOnline();
+
+       // Configure the things attached to the core
+       List<X509Certificate> clientCertificates = createClientCertificates(1);
+       String clientPem = CertificateHelper.toPem(clientCertificates.get(0));
+       Supplier<String> thingOne =  () -> "ThingOne";
+       iotAuthClientFake.attachCertificateToThing(thingOne.get(), clientPem);
+       iotAuthClientFake.attachThingToCore(thingOne);
+       iotAuthClientFake.activateCert(clientPem);
+
+       runNucleusWithConfig("config.yaml");
+       assertNotNull(connectToCore(thingOne.get(), clientPem));
+
+       // When
+       network.goOffline();
+       iotAuthClientFake.deactivateCert(clientPem); // Revoked
+       assertNotNull(connectToCore(thingOne.get(), clientPem));
+
+       // Then
+       network.goOnline();
+       assertThrows(AuthenticationException.class, () -> {connectToCore(thingOne.get(), clientPem);});
```

Comment:
this is a race, isn't it? Between the online callback and trying `connectToCore`?

Reply:
All this is synchronous for the purpose of this test, but there could be a race between the time we record the state as "online" and the time the device connects to core, causing the code to hit the cache instead of going to the cloud.